### PR TITLE
[DO NOT MERGE] PPO with action-dependent baselines

### PR DESCRIPTION
**Pull request by cathywu**
For review only.

### CODE REVIEW

File: python/ray/rllib/ppo/ppo.py [hunk 1]
```diff
@@ -271,7 +334,7 @@ def _fetch_metrics_from_remote_evaluators(self):
     def _stop(self):
         # workaround for https://github.com/ray-project/ray/issues/1516
         for ev in self.remote_evaluators:
-            ev.__ray_terminate__.remote(ev._ray_actor_id.id())
+            ev.__ray_terminate__.remote()
```
Comment (cathywu) [inline python/ray/rllib/ppo/ppo.py]: Why this change?

File: python/ray/rllib/ppo/loss.py [hunk 2]
```diff
@@ -16,16 +16,12 @@ def __init__(
             self, observation_space, action_space,
             observations, value_targets, advantages, actions,
             prev_logits, prev_vf_preds, logit_dim,
-            kl_coeff, distribution_class, config, sess, registry):
+            kl_coeff, distribution_class, config, sess, registry, ADB=False):
         self.prev_dist = distribution_class(prev_logits)
-        self.shared_model = (config["model"].get("custom_options", {}).
```
Comment (cathywu) [inline python/ray/rllib/ppo/loss.py]: Were these lines removed for simplicity or refactored to be elsewhere?

File: python/ray/rllib/ppo/loss.py [hunk 3]
```diff
@@ -38,50 +34,36 @@ def __init__(
...
                 self.value_function = ModelCatalog.get_model(
-                    registry, observations, 1, vf_config).outputs
+                    registry, input_vf, 1, vf_config).outputs
             self.value_function = tf.reshape(self.value_function, [-1])
 
-        curr_logp = self.curr_dist.logp(actions)
-        prev_logp = self.prev_dist.logp(actions)
+        # Make loss functions.
+        if ADB:
+            self.ratio = tf.div(self.curr_dist.r_matrix(actions),
```
Comment (cathywu) [inline python/ray/rllib/ppo/loss.py]: What is this doing? What is `r_matrix(actions)`?

File: python/ray/rllib/ppo/loss.py [hunk 4]
```diff
@@ -38,50 +34,36 @@ def __init__(
...
-            kl_prod = tf.add_n([kl_coeff[0] * kl_i for
-                                i, kl_i in enumerate(self.kl)])
-            # all the above values have been rescaled by num_agents
-            self.surr /= self.num_agents
-            kl_prod /= self.num_agents
-            entropy_prod /= self.num_agents
+        self.surr1 = self.ratio * advantages
+        if ADB:
+            action_dim = action_space.shape[0]
+            self.surr2 = tf.clip_by_value(self.ratio, (1 - config["clip_param"])**(1/action_dim),
```
Comment (cathywu) [inline python/ray/rllib/ppo/loss.py]: Are these ranges reasonable to you?

For m=3 actions, `clip_param`=0.2:
(0.928, 1.063)

For m=8 actions, `clip_param`=0.2:
(0.972, 1.023)

File: python/ray/rllib/ppo/loss.py [hunk 5]
```diff
@@ -94,15 +76,15 @@ def __init__(
             self.vf_loss2 = tf.square(vf_clipped - value_targets)
             self.vf_loss = tf.minimum(self.vf_loss1, self.vf_loss2)
             self.mean_vf_loss = tf.reduce_mean(self.vf_loss)
-            self.loss = tf.reduce_mean(
-                -self.surr + kl_prod +
-                config["vf_loss_coeff"] * self.vf_loss -
-                entropy_prod)
+            self.loss = tf.reduce_mean(-self.surr)
```
Comment (cathywu) [inline python/ray/rllib/ppo/loss.py]: Why this change?

File: python/ray/rllib/ppo/loss.py [hunk 6]
```diff
@@ -113,11 +95,34 @@ def __init__(
...
+            [self.sampler, self.curr_logits],
             feed_dict={self.observations: [observation]})
-        return action[0], {"vf_preds": vf[0], "logprobs": logprobs[0]}
+        vf = self.sess.run(
+            self.value_function,
+            feed_dict={self.observations: [observation], self.actions: action})
+
+        return action[0], [], {"vf_preds": vf[0], "logprobs": logprobs[0]}
+
+    def compute_Q_fuctions(self, observations, actions):
```
Comment (cathywu) [inline python/ray/rllib/ppo/loss.py]: Does this compute `m` baseline for each sample?

File: python/ray/rllib/ppo/ppo.py [hunk 7]
```diff
@@ -31,9 +34,11 @@
     # Initial coefficient for KL divergence
     "kl_coeff": 0.2,
     # Number of SGD iterations in each outer loop
-    "num_sgd_iter": 30,
+    "num_sgd_iter_policy": 30,
+    "num_sgd_iter_vf": 30,
```
Comment (cathywu) [inline python/ray/rllib/ppo/ppo.py:38]: Number of SGD iterations for fitting the value function? (Or action-value function Q?)

**Review by cathywu — COMMENTED** [review]
Here's a suggestion for a troubleshooting experiment we can do:

Pass in zeros as actions when fitting the Q function and evaluating them to compute the AD baselines. This should reduce the implementation to the value function baseline case (approximately).
